Gems/Atom/Tools/ShaderManagementConsole/Code/Source/Document/ShaderManagementConsoleDocument.cpp

```diff
@@ -215,14 +205,59 @@ namespace ShaderManagementConsole
 
     bool ShaderManagementConsoleDocument::IsModified() const
     {
-        return false;
+        return CanUndo();
+    }
+
+    bool ShaderManagementConsoleDocument::BeginEdit()
+    {
+        // Save the current properties as a momento for undo before any changes are applied
+        m_shaderVariantListSourceDataBeforeEdit = m_shaderVariantListSourceData;
+        return true;
+    }
+
+    bool ShaderManagementConsoleDocument::EndEdit()
+    {
+        bool modified = false;
+
+        // Lazy evaluation, comparing the current and previous shader variant list source data state to determine if we need to record undo/redo history.
+        // TODO Refine this so that only the deltas are stored.
+        const auto& undoState = m_shaderVariantListSourceDataBeforeEdit;
+        const auto& redoState = m_shaderVariantListSourceData;
+        if (undoState.m_shaderFilePath != redoState.m_shaderFilePath ||
+            undoState.m_shaderVariants.size() != redoState.m_shaderVariants.size())
+        {
+            modified = true;
+        }
+        else
+        {
+            for (size_t i = 0; i < redoState.m_shaderVariants.size(); ++i)
+            {
+                if (undoState.m_shaderVariants[i].m_stableId != redoState.m_shaderVariants[i].m_stableId ||
+                    undoState.m_shaderVariants[i].m_options != redoState.m_shaderVariants[i].m_options)
+                {
+                    modified = true;
+                    break;
+                }
+            }
+        }
+
+        if (modified)
+        {
+            AddUndoRedoHistory(
+                [this, undoState]() { SetShaderVariantListSourceData(undoState); },
+                [this, redoState]() { SetShaderVariantListSourceData(redoState); });
```

Comment:
Won't be any worse than what we do in the main editor but I do agree. There is already a to do comment at the top of the function to revisit this and only store the deltas then have undo and redo only apply patches. I did not do it initially because I needed to put extra thought into how to remove entries on an undo operation but that can be done just by storing empty strings as a marker in the recorded data. Worth revisiting and possibly even adding the functions directly to the shader variant list source data.